### ISSUE TITLE
Reusing Components in Multiple Mixtures

### ISSUE THREAD

**Issue by WilliamIX** [BuildACell/BioCRNPyler#123]
Currently, BioCRNpyler cannot consistently reuse Components in multiple Mixtures. What I mean by this is that in cases such as the below:
```
G = DNAassembly("G")
M1 = Mixture(components = [G])
M2 = DifferentMixture(components = [G])
```
M1 will consistently compile correctly, but M2 may not be able to compile correctly. This is because in some cases, the Mixture does things to the component during compilation. For example ExpressionExtract and ExpressionDilutionMixture tell DNAassembly to not transcribe in a subtle way in order to model gene expression without both transcription and translation. 

Goal: Ideally, I think we should be able to reuse Components in multiple Mixtures OR completely not allow this from happening via errors.

I am not sure what the best solution is. One idea is to find a way to call `Component.__init__` at compile time so that any changes done by the Mixture can be undone by re-initializing the component. However - if a person has made changes to the Component before compilation these will then be undone!

Maybe the answer is for ExpressionExtract and ExpressionMixture to work differently - but, in general, I still think we need to safeguard Mixtures from modifying Components or allow for some kind of reset functionality so that the Components can be reused later while preserving the ability of a user to modify Components before Compilation? 

@dr3y do you think there will be issues in this sense with regard to DNAconstruct? I think we should probably build any solution (even if it is "DON'T REUSE COMPONENTS") with that class in mind because it is one of the most complicated in the package.

**Comment by murrayrm** [comment]
One seeming easy possibility would be for mixtures to make copies of any components they are passed.  If you want to modify something for an entire component, do it before you pass to a mixture.  Otherwise, mixture dependent changes only affect that mixture.

Not sure if that matches the DNAconstruct use case.

**Comment by WilliamIX** [comment]
Issue solution:

```
G = DNAassembly("G")
M1 = Mixture(components = [G])
M2 = DifferentMixture(components = [G])
```
In these cases, G will be copied into each mixture. To edit the copy in the Mixture we can say:
`M1[G]` ==> returns the copy of G in M1
OR
`M1.components[i]` returns the ith component of M1
which can then be edited. Furthermore
`M1.add_component(G)` should return the copy of G in M1

**Comment by zoltuz** [comment]
@WilliamIX , was this addressed in PR #134 ?
